packages/core/src/sync-realtime/index.ts

```diff
@@ -807,11 +816,63 @@ export const createRealtimeSync = (
         },
       });
 
-      const enqueue = async () => {
+      const watchWebsocket = async () => {
+        if (wsSubscriptionId === undefined) {
+          // Subscribe only if there is no active websocket connection
+          try {
+            const { subscriptionId, unsubscribe } =
+              await _eth_subscribe_newHeads(args.requestQueue, {
+                onData: async (data) => {
+                  await enqueue(data.result.hash);
+                },
+                onError: async (error) => {
+                  // TODO: handle this error
+                  wsSubscriptionId = undefined;
+                  if (unsubscribe) {
+                    await unsubscribe();
+                  }
+                  throw new Error(error);
+                },
+              });
+
+            consecutiveErrors = 0;
+            wsSubscriptionId = subscriptionId;
+            wsUnsubscribe = unsubscribe;
+          } catch (_error) {
+            if (isKilled) return;
+
+            const error = _error as Error;
+
+            args.common.logger.warn({
+              service: "realtime",
+              msg: `Failed to subscribe to the latest '${args.network.name}' block`,
+              error,
+            });
+
+            // After a certain number of attempts, emit a fatal error.
+            if (++consecutiveErrors === ERROR_TIMEOUT.length) {
+              args.common.logger.error({
+                service: "realtime",
+                msg: `Fatal warning: Failed to subscribe to the latest '${args.network.name}' block after ${ERROR_TIMEOUT.length} attempts.`,
+                error,
+              });
+
+              args.onFatalError(error);
+            }
+          }
+        }
+      };
+
+      const enqueue = async (blockHash?: Hash) => {
         try {
-          const block = await _eth_getBlockByNumber(args.requestQueue, {
-            blockTag: "latest",
-          });
+          const block =
+            blockHash === undefined
+              ? await _eth_getBlockByNumber(args.requestQueue, {
+                  blockTag: "latest",
+                })
+              : await _eth_getBlockByHash(args.requestQueue, {
```

Comment:
how come we still need to fetch the block? Does `newHeads` not return the full block object?

Reply:
`newHeads` returns the block header, which does not include transactions. So, we still need to fetch via _eth_getBlockByHash to access complete block data.